src/contracts/StakingChef.sol

```diff
@@ -95,73 +77,72 @@ contract StakingChef is Chef {
-        harvest(stakeIndex);
+    function claim() public {
+        Staker storage stake = staker[msg.sender];
+        harvest();
 
         require(stake.claimable != 0, "No claimable rewards to withdraw");
-        require(stake.stakeAddress == _msgSender(), "Sender can not claim");
 
         SafeERC20.safeTransfer(IERC20(getMetricToken()), msg.sender, stake.claimable);
         stake.claimable = 0;
 
-        emit Withdraw(msg.sender, stakeIndex, stake.claimable);
+        emit TransferPrincipal(msg.sender, stake, stake.claimable);
     }
 
-    function withdrawPrincipal(uint256 stakeIndex) public {
-        Staker storage stake = _stakes[stakeIndex];
-
+    function withdrawPrincipal() public {
+        Staker storage stake = staker[msg.sender];
         require(stake.metricAmount != 0, "No Metric to withdraw");
-        require(stake.stakeAddress == _msgSender(), "Sender can not withdraw");
 
         SafeERC20.safeTransfer(IERC20(getMetricToken()), msg.sender, stake.metricAmount);
         stake.metricAmount = 0;
 
-        emit Withdraw(msg.sender, stakeIndex, stake.metricAmount);
+        emit TransferPrincipal(msg.sender, stake, stake.metricAmount);
     }
 
-    function harvest(uint256 stakeIndex) internal {
-        Staker storage stake = _stakes[stakeIndex];
-
+    function harvest() internal {
+        Staker storage stake = staker[msg.sender];
         updateAccumulatedStakingRewards();
 
         uint256 claimable = stake.metricAmount.mul(getLifetimeShareValue()).div(ACC_METRIC_PRECISION).sub(stake.rewardDebt);
 
         stake.rewardDebt = claimable;
         stake.claimable = stake.claimable.add(claimable);
-        emit Harvest(msg.sender, stakeIndex, claimable);
+        emit HarvestRewards(msg.sender, stake, claimable);
     }
 
-//------------------------------------------------------Getters
+    //------------------------------------------------------Getters
 
-    function getStakes() public view returns (Staker[] memory) {
-        return _stakes;
-    }
+    // TODO figure out how to view stakes now that the stake array has been removed
+    // function getStakes() public view returns (Staker[] memory) {
+    //     return _stakes;
+    // }
 
-//------------------------------------------------------Distribution
+    //------------------------------------------------------Distribution
 
-    function viewPendingHarvest(uint256 stakeIndex) public view returns (uint256) {
-        Staker storage stake = _stakes[stakeIndex];
+    function viewPendingHarvest() public view returns (uint256) {
+        Staker storage stake = staker[msg.sender];
 
         return stake.metricAmount.mul(getLifetimeShareValue()).div(ACC_METRIC_PRECISION).sub(stake.rewardDebt);
     }
 
-    function viewPendingClaims(uint256 stakeIndex) public view returns (uint256) {
-        Staker storage stake = _stakes[stakeIndex];
+    function viewPendingClaims() public view returns (uint256) {
+        Staker storage stake = staker[msg.sender];
```

Comment:
I'm thinking here you could use memory in place of storage? since its read only?

https://medium.com/coinmonks/ethereum-solidity-memory-vs-storage-which-to-use-in-local-functions-72b593c3703a
referencing getUsingMemory function in this article

Reply:
I think you have to use storage because staker[msg.sender] is already in storage